stablehlo/dialect/StablehloOps.cpp

```diff
@@ -4660,48 +4660,80 @@ LogicalResult ReplicaIdOp::inferReturnTypes(
-                                     /*branchName=*/"false_branch"))) {
+LogicalResult IfOp::inferReturnTypeComponents(
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  IfOp::Adaptor adaptor(operands, attributes, regions);
+  if (failed(verifyConditionalBranch(adaptor.true_branch(),
+                                     /*branchName=*/"true_branch", location)))
     return failure();
-  }
+  if (failed(verifyConditionalBranch(adaptor.false_branch(),
+                                     /*branchName=*/"false_branch", location)))
+    return failure();
+
+  auto trueBranchRetureTypes =
+      adaptor.true_branch().front().getTerminator()->getOperandTypes();
+  auto falseBranchRetureTypes =
+      adaptor.false_branch().front().getTerminator()->getOperandTypes();
+  
+  if (!isCompatibleForTypeRange(trueBranchRetureTypes, falseBranchRetureTypes))
+    return emitOptionalError(
+        location, "true_branch and false_branch mismatched return types: ",
+        trueBranchRetureTypes, " vs ", falseBranchRetureTypes);
+
+  for (auto trueBranchRetureType : trueBranchRetureTypes)
+    inferredReturnShapes.emplace_back(trueBranchRetureType.cast<ShapedType>());
   return success();
 }
 
 //===----------------------------------------------------------------------===//
 // Case Op
 //===----------------------------------------------------------------------===//
 
-LogicalResult CaseOp::verify() {
-  auto numBranches = branches().size();
-
-  for (unsigned i = 0; i < numBranches; ++i)
-    if (failed(verifyConditionalBranch(*this, branches()[i],
-                                       /*branchName=*/"branch " + Twine(i))))
+LogicalResult CaseOp::inferReturnTypeComponents(
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  CaseOp::Adaptor adaptor(operands, attributes, regions);
+  ValueTypeRange<OperandRange> branch0ResultTypes =
+      adaptor.branches()[0]->front().getTerminator()->getOperandTypes();
+
+  for (unsigned i = 0; i < adaptor.branches().size(); ++i) {
+    Twine branchName = "branch " + Twine(i);
+    Region* region = adaptor.branches()[i];
+    if (failed(verifyConditionalBranch(*region, branchName, location)))
       return failure();
 
+    auto branchResultTypes = region->front().getTerminator()->getOperandTypes();
+    if (!isCompatibleForTypeRange(branch0ResultTypes, branchResultTypes))
+      return emitOptionalError(location, "branch 0 and ", branchName,
+                               " have mismatched return types: ",
+                               branch0ResultTypes, " vs ", branchResultTypes);
+  }
+  for (auto resultType : branch0ResultTypes)
+    inferredReturnShapes.emplace_back(resultType.cast<ShapedType>());
   return success();
 }
```

Comment:
I think the body of this function should be factored out into a helper function called from both IfOp and CaseOp type inference, to deduplicate the existing implementation. Something like:

```
static TypeRange inferConditionalReturnTypeComponents(Optional<Location> location, RegionRange branches) {
  // 1) Verify that branches is non-empty.
  // 2) Pick the type of the first branch.
  // For each branch:
  // 3) Verify that the branch has zero arguments.
  // 4) Verify that it has a compatible type.
}
```

`verifyConditionalBranch` could be merged into this function - I don't think it's needed as a separate thing in addition to `inferConditionalReturnTypeComponents`.

Reply:
Actually I was thinking of similar way and what's stop me was considering that I need pass in a different list of branch names like `branch_true` or `branch 0`, and need make sure the len(branch_names) == regionRange.size, or use Array<pair<branch, branch_name>>, which is complicated. 
But now I prefer your suggested way with general branch names for both IfOp and CaseOp: branch 0, branch 1 ... Because now I feel these also good names for IfOp too: in mlir test, true and false branches exists in similar way like CaseOp.
Fixed. And now almost all code are shared!

Reply:
Very nice! 👏